File: src/algorithms/sorting/exchange/gnome-sort/sources/gnome-sort.ts
// Gnome Sort — move forward if in order, backward (swapping) if not
function gnomeSort(inputArray: number[]): number[] {
  // @step:initialize
  const sortedArray = [...inputArray]; // @step:initialize
  const arrayLength = sortedArray.length; // @step:initialize
  let position = 0; // @step:initialize

  while (position < arrayLength) {
    if (position === 0) {
      // @step:move-forward
      position++; // @step:move-forward
    } else {
      // @step:compare
      if (sortedArray[position]! >= sortedArray[position - 1]!) {
        // Elements are in order — move forward
        // @step:move-forward
        position++; // @step:move-forward
      } else {
        // Elements are out of order — swap and step back
        // @step:swap
        const temporaryValue = sortedArray[position]!; // @step:swap
        sortedArray[position] = sortedArray[position - 1]!; // @step:swap
        sortedArray[position - 1] = temporaryValue; // @step:swap
        position--; // @step:swap
      }
    }
  }

  // All elements are in their sorted positions
  // @step:mark-sorted

  return sortedArray; // @step:complete
}

File: src/algorithms/sorting/exchange/gnome-sort/sources/gnome-sort.py
def gnome_sort(input_array: list[int]) -> list[int]:  # @step:initialize
    sorted_array = input_array.copy()  # @step:initialize
    array_length = len(sorted_array)  # @step:initialize
    position = 0  # @step:initialize

    while position < array_length:
        if position == 0:
            # @step:move-forward
            position += 1  # @step:move-forward
        else:
            # @step:compare
            if sorted_array[position] >= sorted_array[position - 1]:  # @step:compare
                # Elements are in order — move forward
                # @step:move-forward
                position += 1  # @step:move-forward
            else:
                # Elements are out of order — swap and step back
                # @step:swap
                temporary_value = sorted_array[position]  # @step:swap
                sorted_array[position] = sorted_array[position - 1]  # @step:swap
                sorted_array[position - 1] = temporary_value  # @step:swap
                position -= 1  # @step:swap

    # All elements are in their sorted positions
    # @step:mark-sorted

    return sorted_array  # @step:complete
